Update existing resources when CR changes

jianrongzhang89 opened the pull request:
## Description
Update the deployed resources when backstage CR is updated.

## Which issue(s) does this PR fix or relate to
- Fixed #72 

## PR acceptance criteria
- [X ] Tests
- [ ] Documentation

## How to test changes / Special notes to the reviewer
1) Deploy the operator on OpenShift and examples/bs1.yaml.
2) Update the CR spec to include:
```
spec:
   application:
      replicas: 2
```
and verify that the deployed backstage has 2 pods running.
3) Update the CR spec to include:
```
spec:
   application:
      extraEnvs:
        envs:
        - name: myenv1
          value: my-value-1
        - name: myenv2
          value: my-value-2
```
and check that the deployed backstage pod has env vars myenv1 and myenv2 set in the main container.
4) Update the CR spec to disable the route:
```
spec:
   application:
      route:
        enabled: true
```
and verify that the backstage route bas been cleaned up.
5)  Update the CR spec to disable the localDB:
```
spec:
  database:
    enableLocalDb: false
    authSecretName:  existing-secret
```
and verify that the local db statefulset, services and secret that were deployed earlier are now deleted.

Note that two status condition types "RouteSynced" and "LocalDbSynced" are introduced to support the functionality required.

rm3l commented on api/v1alpha1/backstage_types.go:
```diff
@@ -21,6 +21,11 @@ import (
 	RuntimeConditionSynced  string = "RuntimeSyncedWithConfig"
+	RouteSynced             string = "RouteSynced"
+	LocalDbSynced           string = "LocalDbSynced"
+	SynckOK                 string = "SyncOK"
```
Typo?

```suggestion
	SyncOK                 string = "SyncOK"
```

rm3l commented on api/v1alpha1/backstage_types.go:
```diff
@@ -21,6 +21,11 @@ import (
+	RouteSynced             string = "RouteSynced"
+	LocalDbSynced           string = "LocalDbSynced"
+	SynckOK                 string = "SyncOK"
+	SynckFailed             string = "SyncFailed"
```
Typo?

```suggestion
	SyncFailed             string = "SyncFailed"
```

jianrongzhang89 replied:
Corrected

rm3l commented on bundle/manifests/backstage-operator.clusterserviceversion.yaml:
```diff
@@ -206,7 +206,7 @@ spec:
                 - name: RELATED_IMAGE_backstage
                   value: quay.io/janus-idp/backstage-showcase:next
-                image: quay.io/janus/operator:next
+                image: quay.io/janus/operator:v0.0.1
```
This tag does not exist.

```suggestion
                image: quay.io/janus/operator:next
```

jianrongzhang89 replied:
Good. Done.

rm3l reviewed (commented):
Seems to work fine at first sight, but I also noticed a slight issue:

`imagePullSecrets` seem to keep adding to the Backstage Deployment Spec if CR contains a `spec.application.imagePullSecrets` and is updated (I just updated the image and number of replicas), e.g.:

```
$ kubectl get deployment backstage-my-rhdh -o yaml
...
       volumeMounts:                                                                                                                                                                          
        - mountPath: /opt/app-root/src/dynamic-plugins-root                                                                                                                                    
          name: dynamic-plugins-root                                                                                                                                                           
        - mountPath: /opt/app-root/src/app-config-rhdh.yaml                                                                                                                                    
          name: app-config-rhdh                                                                                                                                                                
          subPath: app-config-rhdh.yaml                                                                                                                                                        
      dnsPolicy: ClusterFirst                                                                                                                                                                  
      imagePullSecrets:                                                                                                                                                                        
      - name: rhdh-pull-secret                                                                                                                                                                 
      - name: rhdh-pull-secret                                                                                                                                                                 
      - name: rhdh-pull-secret
      - name: rhdh-pull-secret
      - name: rhdh-pull-secret
      - name: rhdh-pull-secret
      - name: rhdh-pull-secret
      - name: rhdh-pull-secret
      - name: rhdh-pull-secret
      - name: rhdh-pull-secret
      - name: rhdh-pull-secret
      - name: rhdh-pull-secret
      - name: rhdh-pull-secret
      - name: rhdh-pull-secret
      - name: rhdh-pull-secret
      - name: rhdh-pull-secret
      - name: rhdh-pull-secret
      - name: rhdh-pull-secret
      - name: rhdh-pull-secret
      - name: rhdh-pull-secret
...
```

Can you take a look?

jianrongzhang89 commented:
@rm3l fixed imagePullSecrets issue.

gazarenkov commented on controllers/backstage_controller.go:
```diff
@@ -108,41 +119,44 @@ func (r *BackstageReconciler) Reconcile(ctx context.Context, req ctrl.Request) (
 		}
-
+		setStatusCondition(&backstage, bs.LocalDbSynced, v1.ConditionTrue, bs.SyncOK, "")
+	} else if isLocalDbDeployed(backstage) { // EnableLocalDb is off but local db has been deployed. Clean up the deployed local db resources
```
I do not think using CR.status field (which is inside isLocalDbDeployed) is legitimate for K8s.
I believe you should use an actual object state comparing it with desired one in every reconciliation.

jianrongzhang89 replied:
See above.

rm3l commented on controllers/backstage_deployment.go:
```diff
@@ -312,3 +286,29 @@ func (r *BackstageReconciler) setDefaultDeploymentImage(deployment *appsv1.Deplo
+			})
+		}
+
+		if len(backstage.Spec.Application.ImagePullSecrets) > 0 { // use image pull secrets from the CR spec
```
What would happen if `backstage.Spec.Application.ImagePullSecrets` was non-empty, and then updated to an empty list? Shouldn't we update it in the Deployment as well?
I noticed that the Deployment still had the previous pull secret..

jianrongzhang89 replied:
Updated the PR so that in this case the deployment should be updated. Also as discussed, changed backstage.Spec.Application.ImagePullSecrets to a pointer so that we can distinguish between the default  (the pointer not set) and empty values for ImagePullSecrets.

jianrongzhang89 commented on api/v1alpha1/backstage_types.go:
```diff
@@ -21,6 +21,11 @@ import (
 const (
 	RuntimeConditionRunning string = "RuntimeRunning"
 	RuntimeConditionSynced  string = "RuntimeSyncedWithConfig"
+	RouteSynced             string = "RouteSynced"
```
This new status is a technical one that allows the reconciler to determine if the route has been deployed earlier.
We will address remaining issues regarding status in a follow up PR.

jianrongzhang89 commented on controllers/backstage_route.go:
```diff
@@ -24,50 +24,64 @@ import (
 	name := fmt.Sprintf("backstage-%s", backstage.Name)
-	route.Name = name
-	route.Spec.To.Name = name
+	route := &openshift.Route{
```
The same way as other objects. Handled in the routeObjectMutFun function.

jianrongzhang89 commented on controllers/backstage_route.go:
```diff
@@ -24,50 +24,64 @@ import (
-		if !errors.IsNotFound(err) {
-			return fmt.Errorf("failed to get backstage route, reason: %s", err)
+	if !shouldCreateRoute(*backstage) {
+		if isRouteDeployed(*backstage) {
```
Well, CR.status is part of  K8s CRD indeed.

jianrongzhang89 commented on controllers/backstage_controller_test.go:
```diff
@@ -1370,7 +1527,7 @@ plugins: []
+			Expect(err.Error()).Should(ContainSubstring("authSecretName is required if enableLocalDb is false"))
 		})
 	})
 })
```
Lets handle such enhancement as part of issue #120.

jianrongzhang89 commented on config/manager/kustomization.yaml:
```diff
@@ -5,7 +5,7 @@ kind: Kustomization
 - name: controller
   newName: quay.io/janus/operator
-  newTag: v0.0.1
+  newTag: next
```
This is requested by @rm3l. Lets discuss tomorrow.

jianrongzhang89 replied:
This change was requested by @rm3l. Could you clarify why it was needed @rm3l ?

gazarenkov commented on controllers/backstage_controller.go:
```diff
@@ -108,41 +119,44 @@ func (r *BackstageReconciler) Reconcile(ctx context.Context, req ctrl.Request) (
+		setStatusCondition(&backstage, bs.LocalDbSynced, v1.ConditionTrue, bs.Deleted, "")
 	}
 
-	err := r.applyBackstageDeployment(ctx, backstage, req.Namespace)
```
I see no contradictions TBH, Reconcile calls applyX (and many others too) therefore it is not one-time?

gazarenkov reviewed (changes requested):
The main reason we postpone Update to the later version is edge cases which we have due to our configuration specific. I think it would be worth to investigate these cases  before and come to some conclusions.
I've created some some issues for that, please [take a look](https://github.com/janus-idp/operator/issues/124), it would be good to resolve it as a prerequisite.

jianrongzhang89 commented on controllers/backstage_route.go:
```diff
@@ -24,50 +24,67 @@ import (
-	if err != nil {
-		if !errors.IsNotFound(err) {
-			return fmt.Errorf("failed to get backstage route, reason: %s", err)
+	if _, err := controllerutil.CreateOrUpdate(ctx, r.Client, route, r.routeObjectMutFun(ctx, route, *backstage, ns)); err != nil {
```
controllerutil.CreateOrUpdate() is a nice utility library function call that handles the logic like get the current resource and merge it with the new spec using the mutate(), and then do the update.

jianrongzhang89 commented on controllers/backstage_deployment.go:
```diff
@@ -312,3 +288,30 @@ func (r *BackstageReconciler) setDefaultDeploymentImage(deployment *appsv1.Deplo
+			})
+		}
+		if backstage.Spec.Application.ImagePullSecrets != nil { // use image pull secrets from the CR spec
+			deployment.Spec.Template.Spec.ImagePullSecrets = nil
```
This is because deployment.Spec.Template.Spec.ImagePullSecrets and backstage.Spec.Application.ImagePullSecrets have different data types.

jianrongzhang89 replied:
Since in other places for configmaps and secrets we chose to use our own types (string) instead of the standard one, I think it makes sense we use the string type here for consistency, or we may want to change all such fields to use the standard one instead.

jianrongzhang89 commented:
> The main reason we postpone Update to the later version is edge cases which we have due to our configuration specific. I think it would be worth to investigate these cases before and come to some conclusions. I've created some some issues for that, please [take a look](https://github.com/janus-idp/operator/issues/124), it would be good to resolve it as a prerequisite.
 
Yes resource update involves various use cases as you pointed out. Most of the ones in #124 are for default config/configmao/secret changes etc and they can be handled separately later. This PR is for the application of the CR changes only.

jianrongzhang89 commented on controllers/backstage_service.go:
```diff
@@ -23,46 +23,64 @@ import (
 		}
-	}
+		service.Spec.ClusterIP = targetService.Spec.ClusterIP
+		for _, ip1 := range targetService.Spec.ClusterIPs {
```
When a service is created, normally k8s may update the spec to include the clusterIP and clusterIPs assigned randomly. These values can not be updated later.

gazarenkov replied:
Well, it is often use case but not only, user is able to specify it or to set None for headless service
https://kubernetes.io/docs/concepts/services-networking/service/

jianrongzhang89 replied:
Yes the user is able to create headless service (actually the operator deploys a headless service for local db). What I meant here after the service has been created, the cluserIP should not be updated.

gazarenkov replied:
Not sure I understand it.
Why we have this opinion and check it ourselves?

jianrongzhang89 replied:
Because without this checking, reconciliation would fail, as the operator tries to override the IPs that were populated by k8s, leading to an update error.

gazarenkov replied:
It is not valid, I believe 
According to https://kubernetes.io/docs/concepts/services-networking/service/
If you do not specify ClusterIP (which is a case a believe) it is is assigned by K8s, i.e it is controlled be Kubernetes, Operator's desired state is still "I let K8s manage it" so, I believe operator must not try to override it.

gazarenkov replied:
I do not think current implementation of update is correct, as discussed.

In general, an Operator must not deal with spec fields
"Types with both spec and status stanzas can (and usually should) have distinct authorization scopes for them. This allows users to be granted full write access to spec and read-only access to status, while relevant controllers are granted read-only access to spec but full write access to status."

And in this particular case, as stated https://kubernetes.io/docs/concepts/services-networking/cluster-ip-allocation/ if ClusterIP is dynamically assigned by control plane (the case we are testing) we should not manage it during CR update (specified desired state is ClusterIP=="", so it managed dynamically by K8s Control Plane).

So, I believe we have to revert this PR and rework this functionality.

jianrongzhang89 commented on controllers/local_db_services.go:
```diff
@@ -63,51 +61,65 @@ import (
+		if len(targetService.Spec.ClusterIP) > 0 && service.Spec.ClusterIP != "" && service.Spec.ClusterIP != "None" && service.Spec.ClusterIP != targetService.Spec.ClusterIP {
+			return fmt.Errorf("db service IP can not be updated: %s, %s, %s", targetService.Name, targetService.Spec.ClusterIP, service.Spec.ClusterIP)
+		}
+		service.Spec.ClusterIP = targetService.Spec.ClusterIP
```
same as above

gazarenkov replied:
same as above

jianrongzhang89 commented on api/v1alpha1/backstage_types.go:
```diff
@@ -21,6 +21,11 @@ import (
+	RouteSynced             string = "RouteSynced"
+	LocalDbSynced           string = "LocalDbSynced"
+	SyncOK                  string = "SyncOK"
+	SyncFailed              string = "SyncFailed"
```
See above.

gazarenkov commented on api/v1alpha1/backstage_types.go:
```diff
@@ -21,6 +21,11 @@ import (
 	RuntimeConditionSynced  string = "RuntimeSyncedWithConfig"
+	RouteSynced             string = "RouteSynced"
+	LocalDbSynced           string = "LocalDbSynced"
+	SyncOK                  string = "SyncOK"
```
I mean RuntimeSyncedWithConfig looks like the same thing as SyncOK, no?

rm3l commented on Makefile:
```diff
@@ -322,7 +322,7 @@ endif
 # https://github.com/operator-framework/community-operators/blob/7f1438c/docs/packaging-operator.md#updating-your-existing-operator
 .PHONY: catalog-build
-catalog-build: opm bundle-push ## Generate operator-catalog dockerfile using the operator-bundle image built and published above; then build catalog image
+catalog-build: bundle-push opm ## Generate operator-catalog dockerfile using the operator-bundle image built and published above; then build catalog image
```
Why is the order changed? I think we need `opm` target first, as it is invoked below

```suggestion
catalog-build: opm bundle-push ## Generate operator-catalog dockerfile using the operator-bundle image built and published above; then build catalog image
```

jianrongzhang89 replied:
The bundle image must be pushed first before catalog image cab be built. opm installation is also required. The order does not matter. I switched them as it is more natural to start the catalog build right after opm is installed.

rm3l replied:
Okay - thanks.

jianrongzhang89 replied:
Yes this is expected, because the host is already set. "subdomain is a DNS subdomain that is requested within the ingress controller's domain (as a subdomain). If host is set this field is ignored."
https://github.com/openshift/api/blob/master/route/v1/types.go#L99

rm3l reviewed (approved):
I've tested this PR against scenarios depicted in https://github.com/janus-idp/operator/blob/main/.rhdh/docs/openshift.adoc, and everything works fine.
We'll need to update https://github.com/janus-idp/operator/blob/main/.rhdh/docs/openshift.adoc (to remove the need to manually delete the Backstage Deployment upon CR changes), but this can done in a subsequent PR/issue.
I think we can go ahead and merge this PR, so that QE can start testing, and also to simplify the docs workflow after CR is modified by the user.
#70 will also need to take that into account.

jianrongzhang89 commented:
> Should it be possible to update the route sub-domain after it is created? I first created a CR with no `spec.application.route` field set. Then I added `spec.application.route.subdomain` to the CR and updated it. But I noticed that the Route was not updated. Is that expected?

Yes this is expected. Since the route host was already created, the subdomain change in the route is ignored by the OpenShift server.
